verilog/tools/kythe/kythe_facts_extractor.cc

```diff
@@ -0,0 +1,205 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <iostream>
+#include <string>
+
+#include "absl/strings/escaping.h"
+#include "absl/strings/substitute.h"
+#include "verilog/tools/kythe/kythe_facts_extractor.h"
+#include "verilog/tools/kythe/kythe_schema_constants.h"
+
+namespace verilog {
+namespace kythe {
+
+void KytheFactsExtractor::Visit(const IndexingFactNode& node) {
+  const auto tag =
+      static_cast<IndexingFactType>(node.Value().GetIndexingFactType());
+
+  VName vname(file_path_);
+
+  switch (tag) {
+    case IndexingFactType::kFile: {
+      vname = ExtractFileFact(node);
+      break;
+    }
+    case IndexingFactType::kModule: {
+      vname = ExtractModuleFact(node);
+      break;
+    }
+    case IndexingFactType::kModuleInstance: {
+      vname = ExtractModuleInstanceFact(node);
+      break;
+    }
+    case IndexingFactType::kVariableDefinition: {
+      vname = ExtractVariableDefinition(node);
+      break;
+    }
+    case IndexingFactType::kVariableReference: {
+      vname = ExtractVariableReference(node);
+      break;
+    }
+  }
+
+  if (tag != IndexingFactType::kFile) {
+    *stream_ << Edge(vname, kEdgeChildOf, vnames_context_.top());
+  }
+
+  const VNameContext::AutoPop vnames_auto_pop(&vnames_context_, &vname);
+  for (const verible::VectorTree<IndexingNodeData>& child : node.Children()) {
+    Visit(child);
+  }
```

Comment:
[no change needed]
Too bad I don't have a variant of `VectorTree::ApplyPreOrder()` that already keeps a context stack for you -- technically each node has access to its ancestors through `VectorTree::Parent()` so a stack would just be redundant and convenient, but I think you want the random access to any ancestor(?).

Reply:
For every line the reachable scope for that line is everything above it inside its scope and anything above it in the outer scope and so on.
e.g.

module m(input x);
endmodule: m;

module b(input y);
//  reference for anything
endmodule: b

The scope for the line with reference will be:
// Global scope
{
  module m,
  module b,
}
// Module b Scope
{
  input y,
}

We can say that the scopes are created on the fly and the tree structure of the scopes is flattened to be linear instead of nested and this is for simplicity and making it possible to use AutoPop.

Reply:
I see.  Don't forget that we also have byte-offset information tracked in many places too.
Verilog has some weird scoping rules in a few places that we want to watch out for in the future, so we can't always rely on location in general for everything.

Reply:
Can u give me some example for that?